### PR TITLE
Utility for writing predictions of selected model groups 

### PR DESCRIPTION
**Pull request by KasunAmare**
This PR contains a script for saving the predictions of a set of selected model groups. This is meant to be used for the following use-case:
- Experiment run with `save_predictions=False` (usually with a large model grid)
- Audition used to narrow down the model groups for post-modeling

The config file for this util takes the experiment hash, a list of model groups, the project path, and optionally a range of `train_end_times`.

### CODE REVIEW

File: src/triage/component/postmodeling/utils/config_file_example.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+# Model group ids we need predictions for
+# List of integers
+model_group_ids:
+  - 1
+  - 2
+
+# This is optional. 
+# If not specificied, all the models in the group will be scored
+# Should be a list of two items
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/config_file_example.yaml:15]: Should probably specify whether these end points will be included in the range.

File: src/triage/component/postmodeling/utils/config_file_example.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# The relevant experiment hash
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/config_file_example.yaml:1]: Might want to name this file something like `add_predictions_example_config.yaml` in case we end creating other utility functions with different config files.

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            This directly writes to the test_results.predictions table
+    """
+    model_matrix_info = _fetch_relevant_model_matrix_info(
+        db_engine=db_engine,
+        experiment_hash=experiment_hash,
+        model_groups=model_groups
+    )
+
+    if len(model_matrix_info)==0:
+        raise ValueError('No models were found for the given experiment and model group(s)')
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:71]: might want to at least generate a warning (or error) if there were any model groups for which no models were found (not just overall)

**Comment by shaycrk** [comment]
Looks good overall, but put a couple of quick notes inline. Would probably also be good to include a brief README.md here with usage instructions.

File: src/triage/component/postmodeling/utils/config_file_example.yaml (hunk 1 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/config_file_example.yaml:15]: yep, added.

File: src/triage/component/postmodeling/utils/config_file_example.yaml (hunk 2 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/config_file_example.yaml:1]: makes sense

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 3 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:65]: Yeah. That's a good point. Set it up so that it will raise an error if any of the model groups are not present. To make sure that the choices in the config file exactly match the saved predictions.

**Comment by KasunAmare** [comment]
Thanks for taking a look. Made the changes and added a README

**Comment by thcrock** [comment]
Should this be in the triage CLI (at src/triage/cli.py) ? It should make for easier discovery. Audition and other postmodeling work already have interfaces there.

**Comment by KasunAmare** [comment]
@thcrock that makes sense to me. @shaycrk what do you think? 

Making it more visible would help a user to plan the experiment better, I guess. As it would enable them to run a larger model grid with `save_predictions=False` without worrying too much about DB space and speed issues brought about by bloating prediction tables.

**Comment by shaycrk** [comment]
@thcrock -- this actually touches on a conversation @KasunAmare and I have been having separately about how some of these pieces fit together. It feels like adding predictions to existing models, adding individual explanations, and predicting forward are all cases where we might want to take one (or a handful) of existing `model_group_id`s and generate new outputs, and might benefit from re-using some of the `experiment` machinery and tracking to do so. I feel like this PR gives us a good starting point, but might change a fair bit if we couple the code more closely with those other use cases as well as existing experiments. If that direction makes sense to you, do you think it makes sense to add a CLI interface now or wait until that's a bit more built out?

**Comment by thcrock** [comment]
This code is adding a command line interface already. My suggestion is just that if we are adding a CLI for a new piece of code, to make it accessible from the triage command instead of fragmenting the CLIs. 

I have no problem with adding new code sans CLIs if the goal is to iterate on it first.

**Comment by thcrock** [comment]
I'm partial to the latter. This code can very easily be ported to the Triage CLI, since it's already using argparse.  I think the main reason to not add a CLI is if you're specifically trying to hide it for now.

**Comment by thcrock** [comment]
And actually, I think you can cut down on a lot of the code in `run` by making use of what's inherited from the root CLI command. You can just pass in the 'dburl' from the root CLI command to `run` and remove all of the dbfile parsing code (you could also pass in a db_engine much like the Crosstabs command does). This has the other advantage of not forcing the user to have a db credentials file; the CLI allows them to use environment variables which are safer. This type of standardization is one of the reasons for the Triage CLI vs fragmented smaller ones.

**Comment by shaycrk** [comment]
Yeah, I think @thcrock's points seem reasonable to me, and agree it's useful in the current form as a starting point that can hopefully merge more into the existing experiment (as well as new predict forward) infrastructure.

I think in using it we've also run into a few other small things that might be worth adjusting before merging:
- We probably don't want both an experiment hash and list of model group ids to be required here. Seems like the model groups should be required, and an optional list of experiment hashes to filter down the specific models
- I might be misremembering, but I think we encountered a bug if only one model group was given?
- Might be better to make the date parameters separate arguments, both optional, to allow for specifying open-ended ranges

**Comment by KasunAmare** [comment]
@shaycrk @thcrock -- Made the following changes to the code. 

- made the experiment hash(es) an optional argument. Now, the user can provide a list of model groups and optionally provide experiment hashes, and/or train end times to restrict the model ids to be scored
- made the train end time rages two separate arguments allowing open-ended intervals
- fixed the bug with the single model group
- merged this with the triage CLI. As Tristan pointed out, this enabled completely removing the `run` method, loading configs,  and db engine creation. 
- Added a method to produce any info/warning messages if there were special characteristics in the models being scored, e.g., multiple model ids for the same model group for the same train end time.

Sorry, this took so long. If you guys can take a look that'll be great. 

Question -- I am currently testing this with the ACLU database and its experiments. I wasn't sure how to add a test for this in `test_cli.py` with an example config in `example/config`. I'm guessing putting ACLU experiment hashes, model group ids as an example wouldn't be ideal? What would be the best way to add a test?

File: src/triage/cli.py [hunk 4]
```diff
@@ -436,5 +437,32 @@ def history(self, args):
         db_history(dburl=self.root.db_url)
 
 
+@Triage.register
+class SavePredictions(Command):
```
Comment (shaycrk) [inline src/triage/cli.py:441]: Might suggest renaming this to avoid confusion since there's a separate CLI argument `save-predictions` that can be used when running the experiment. Perhaps something like `AddPredictions` or something similar?

File: src/triage/component/postmodeling/utils/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+config file example: `add_predictions_example_config.py`
+
+This utility accepts a list of model group ids and an experiment hash, and generates & saves test predictions for the relevant models. This util is primarily targetted to be used post-audition for the following use-case:
+1. A large model grid was used in the experiment with `save_predictions=False`
+2. Audition was used to narrow down model_groups of interest & predictions are required for postmodeling
+
+**Usage**
+
+This util is available with the triage CLI and can be run using the following command
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/README.md:13]: Would also be good to include usage instructions for importing the function and running in a python script separate from the CLI.

File: src/triage/cli.py [hunk 6]
```diff
@@ -436,5 +437,32 @@ def history(self, args):
...
+    """Save the predictions of selected model groups"""
```
Comment (shaycrk) [inline src/triage/cli.py:442]: This just saves test set (not training set) predictions, right? It's pretty rare we use the training set predictions, but might be good to specify this in the docstring/usage docs

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    msk = time_cts > 1
+
+    if len(time_cts[msk]) > 1:
+        logging.warning('There are model groups with more than one model per train_end_time. See below: \n {}'.format(time_cts[msk].reset_index()))
+        
+    
+def generate_predictions(db_engine, model_groups, project_path, experiment_hashes=None, train_end_times_range=None, rank_order='worst', replace=False):
+    """ Generate predictions and write to DB for a set of model_groups in an experiment
+        Args:
+            db_conn: Sqlalchemy engine
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:75]: should be `db_engine:` not `db_conn:`

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            model_groups (list): The list of model group ids we are interested in (ideally, chosen through audition)
+            project_path (str): Path where the created matrices and trained model objects are stored for the experiment
+            experiment_hashes (List[str]): Optional. hash of the experiment (currently handling only one experiment)
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:78]: Didn't you update it to all for more than one experiment?

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    model_matrix_info = _fetch_relevant_model_matrix_info(
+        db_engine=db_engine,
+        model_groups=model_groups,
+        experiment_hashes=experiment_hashes
+    )
+
+    if len(model_matrix_info)==0:
+        raise ValueError('No models were found for the given experiment and model group(s)')
+    
+    # All the model groups specified in the
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:97]: sentence fragment

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    if train_end_times_range is not None: 
+        if 'range_start_date' in train_end_times_range:
+            range_st = train_end_times_range['range_start_date']
+            msk = (model_matrix_info['train_end_time'] >= range_st)
+            logging.info('Filtering out models with a train_end_time before {}'.format(range_st))
+
+            model_matrix_info = model_matrix_info[msk]
+
+        if 'range_end_date' in train_end_times_range:       
+            range_en = train_end_times_range['range_end_date']
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:115]: Why `range_en` not `range_end`?

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            msk = (model_matrix_info['train_end_time'] <= range_en)
+            logging.info('Filtering out models with a train_end_time after {}'.format(range_en))
+
+            model_matrix_info = model_matrix_info[msk]
+
+        if len(model_matrix_info) == 0:
+            raise ValueError('No models were found for the given time range')
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:122]: Maybe consolidate with the check above and only do this once after doing all the filtering?

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    # Sometimes there are more than one model for each model group with different random seeds
+    time_cts = models_df.groupby(['train_end_time', 'model_group_id']).count()['model_id']
+    time_cts.name = 'num_models'
+    msk = time_cts > 1
+
+    if len(time_cts[msk]) > 1:
+        logging.warning('There are model groups with more than one model per train_end_time. See below: \n {}'.format(time_cts[msk].reset_index()))
+        
+    
+def generate_predictions(db_engine, model_groups, project_path, experiment_hashes=None, train_end_times_range=None, rank_order='worst', replace=False):
```
Comment (shaycrk) [inline src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:72]: Does `replace=False` feel like the right default here (especially given that it's currently not passed through from the CLI or config)? For instance, if someone ran an experiment with predictions then fixed the data and re-ran the experiment without predictions, they wouldn't be able to use the CLI to update the predictions after the fact. @thcrock -- what do you think here?

**Comment by shaycrk** [comment]
Thanks @KasunAmare -- I took a pass through and added a few notes inline!

On the question about tests, I think most of the tests in `test_cli.py` simply check that the right functionality is invoked without running it, so you can probably follow the patterns there for the same thing.

To test the `generate_predictions` function itself here, you'll probably want to create a new file in `postmodeling_tests` and then use a test experiment/DB to make sure predictions get added. See [conftest.py](https://github.com/dssg/triage/blob/cc48af48eb06997018887159abe01af8e9aea191/src/tests/conftest.py#L92) for an example of generating such an experiment and then [postmodeling_tests/test_without_predictions.py](https://github.com/dssg/triage/blob/cc48af48eb06997018887159abe01af8e9aea191/src/tests/postmodeling_tests/test_without_predictions.py) for how this can be used.

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 12 above)
Comment (thcrock) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py:72]: Yeah for this task, it seems like replace=True is what I would go with since it's on smaller amounts of data. replace=False seems more useful at Experiment-scale but I'm guessing for something this small I'd just always want it to do the work I'm telling it to do.

File: src/triage/cli.py (hunk 6 above)
Comment (KasunAmare) [reply src/triage/cli.py]: Changed the name and specified that it's only the test predictions

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 7 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py]: fixed

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 8 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py]: Yes, updated the docstring

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 9 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py]: Fixed

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 10 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py]: fixed

File: src/triage/component/postmodeling/utils/predictions_selected_model_groups.py (hunk 12 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/predictions_selected_model_groups.py]: That makes sense. Changed it

File: src/triage/component/postmodeling/utils/README.md (hunk 5 above)
Comment (KasunAmare) [reply src/triage/component/postmodeling/utils/README.md:15]: added

**Review by shaycrk — APPROVED** [review]
Thanks @KasunAmare -- looks good to me! Feel free to merge now if you want (otherwise, I'll merge it along with the other open PRs when we're ready to tag the new release)

**Comment by KasunAmare** [comment]
Thanks, @shaycrk. I'm happy to wait till we tag the new release.
